tests/integration/test_charm_with_loki.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
```

Comment:
as a general comment, consider moving all functions that are not actual `test_`'s to a separate file to make it an easier read. also, instead of having "hidden" (or embedded) asserts, consider making them return booleans instead and assert those from the **actual** test cases.

Reply:
I don't think it's a bad idea to write higher-level assertions to avoid repetitions and to have better stack traces. I'd rather not have pytest tell me every time `False != True` instead of seeing immediately from the stack trace that the error was raised from a `check_that_loki_can_be_reached` or something similarly telling.
It's a pattern I've seen before, but I agree that it can be stripped from simple helpers which only do a single assert.

Reply:
I think it clouds the behavior, and even though I've also seen it in the wild, it's still very much a code smell to me rather than a pattern. Returning the values you are asserting and doing the assertion in the main branch would be equally preferable in my opinion. but it's a minor and I can live with it being as-is as well.